File: core/story_settings.py
"""Per-story gameplay settings and legacy-save normalization."""

from copy import deepcopy

from core.story_length import DEFAULT_TARGET_STORY_LENGTH, normalize_target_story_length


DEFAULT_STORY_SETTINGS = {
    "showFutures": True,
    "showDice": True,
    "allowReroll": True,
    "aiSuggestions": True,
    "entitiesEnabled": True,
    "showEntityNames": False,
    "showEntityMotives": False,
    "allowEntityEditing": False,
    "showEntityBubbles": False,
    "showInfluenceBubbles": False,
    "showCausalLedger": False,
    "showTime": True,
    "autoCompressMemory": False,
    "worldbookCaptureEnabled": True,
    "worldbookCaptureReview": False,
    "learnUserPreferences": True,
    "useUserPreferences": True,
    "deepPreferenceAnalysis": True,
    "analyzeSensitivePreferences": False,
    "preferenceStoryEnabled": True,
    "preferenceAdultEnabled": True,
    "preferenceActionEnabled": True,
    "preferenceCharacterEnabled": True,
    "preferenceRelationshipEnabled": True,
    "preferenceVisualEnabled": True,
    "targetStoryLength": DEFAULT_TARGET_STORY_LENGTH,
}


def normalize_story_settings(settings=None, defaults=None):
    normalized = deepcopy(DEFAULT_STORY_SETTINGS)
    for source in (defaults or {}, settings or {}):
        for key in normalized:
            if key not in source:
                continue
            if key == "targetStoryLength":
                normalized[key] = normalize_target_story_length(source[key])
            else:
                normalized[key] = bool(source[key])
    return normalized

File: aliveworld-ui/src/utils/storySettings.js
export const STORY_SETTING_DEFAULTS = Object.freeze({
  showFutures: true, showDice: true, allowReroll: true, aiSuggestions: true,
  entitiesEnabled: true, showEntityNames: false, showEntityMotives: false,
  allowEntityEditing: false, showEntityBubbles: false, showTime: true,
  showInfluenceBubbles: false,
  showCausalLedger: false,
  autoCompressMemory: false,
  worldbookCaptureEnabled: true,
  worldbookCaptureReview: false,
  learnUserPreferences: true,
  useUserPreferences: true,
  deepPreferenceAnalysis: true,
  analyzeSensitivePreferences: false,
  preferenceStoryEnabled: true,
  preferenceAdultEnabled: true,
  preferenceActionEnabled: true,
  preferenceCharacterEnabled: true,
  preferenceRelationshipEnabled: true,
  preferenceVisualEnabled: true,
  targetStoryLength: 500
});

export function normalizeStorySettings(settings = {}, defaults = STORY_SETTING_DEFAULTS) {
  return Object.fromEntries(Object.keys(STORY_SETTING_DEFAULTS).map(key => {
    const value = settings[key] ?? defaults[key] ?? STORY_SETTING_DEFAULTS[key];
    if (key === 'targetStoryLength') {
      const parsed = Number.parseInt(value, 10);
      return [key, Math.max(200, Math.min(3000, Number.isFinite(parsed) ? parsed : 500))];
    }
    return [key, Boolean(value)];
  }));
}
